Correlate requests and responses by seqid

markerickson-wf opened the pull request:
https://issues.apache.org/jira/browse/THRIFT-3299

This is what I was thinking for correlating requests and responses by seqid.  Let me know what you think.
@evanweible-wf
@stevenosborne-wf

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Add TMessageReader and use in TClientSocketTransport to correlate requests and responses by seqid

markerickson-wf commented:
Note that this works per service Client.  If you wanted to share the same transport between multiple Clients, e.g. LinkingClient and DatatablesClient, then you'd have problems, because they increment the seqid independently.  

You should create a transport per Client in that case, so they keep their completer Maps isolated.

stevenosborne-wf commented:
+1

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Fix typo in error message.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Add support for a timeout when waiting for a response.

markerickson-wf commented:
I added support for a timeout in https://github.com/markerickson-wf/thrift/commit/24fec2938fb43bd264f704b3b5079c9533c7f99a .

evanweible-wf commented on lib/dart/test/transport/t_socket_transport_test.dart:
```diff
@@ -48,6 +52,23 @@ void main() {
+      expect(responseReady, throwsA(new isInstanceOf<TimeoutException>()));
+
+      // allow the timeout to occur
+      await new Future.delayed(new Duration(milliseconds: 100));
```
Is this necessary? I would have thought that line 67 would cover this since `throwsA` should wait for `responseReady` to complete

markerickson-wf replied:
No, it's not necessary - thanks!

evanweible-wf commented on lib/dart/lib/src/transport/t_message_reader.dart:
```diff
@@ -0,0 +1,70 @@
+
+/// [TMessageReader] extracts a [TMessage] from bytes.  This is used to allow a
+/// transport to inspect the message seqid and map responses to requests.
+class TMessageReader extends TTransport {
```
I get that this makes it easy to use a protocol to decode a message since protocols use `TTransport`s, but it does seem wrong to make `TMessageReader` extend `TTransport` given that it doesn't actually send or receive any data and it only supports reading. As of right now, I can't think of an alternative that wouldn't require going down a rabbit hole of refactoring, so I'm alright with this, but something to keep in mind.

markerickson-wf replied:
Good observation.  I moved the transport logic to an internal helper class.

evanweible-wf commented:
+1

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Make TMessageReader not extend TTransport, move to internal class _TMessageReaderTransport

markerickson-wf commented:
@evanweible-wf I addressed your comments in https://github.com/markerickson-wf/thrift/commit/f07181fd866aca1a4ebfdf1cf378a8630fe22100 .

evanweible-wf commented:
+1 looks good!
